assets/dns/daemonset.yaml

```diff
@@ -85,8 +87,11 @@ spec:
   updateStrategy:
     type: RollingUpdate
     rollingUpdate:
-      # TODO: Consider setting maxSurge to a positive value.
-      maxSurge: 0
-      # Note: The daemon controller rounds the percentage up
-      # (unlike the deployment controller, which rounds down).
-      maxUnavailable: 10%
+      # Set maxSurge to a positive value so that each node that has a pod
+      # continues to have a local ready pod during a rolling update.  This is
+      # important for topology-aware hints as well as for similar logic in
+      # openshift-sdn and ovn-kubernetes that prefers to use a local ready DNS
+      # pod whenever one exists.
+      maxSurge: 10%
+      # maxUnavailable must be zero when maxSurge is nonzero.
```

Comment:
*nit* Is "must be zero" true? It is more accurate to say "should be zero"?

I don't see that in the docs, but I agree that it needs to be zero in order to fully use maxSurge, that makes sense. It sounds like you could potentially have both set per the docs, but I'd imagine that'd be an messy update.

Reply:
I agree with @gcs278 .  The docs say maxSurge can not be 0 if MaxUnavailable is 0 and vice versa. Can't they both be non-zero?

Also, 10% seems more like a minimum that we would want, if maxSurge means the maximum number of nodes with an existing available DaemonSet pod that can have an updated DaemonSet pod during during an update.  If we'd want to have a pod available **on every node** before we delete the previous one, wouldn't this need to be 100% ?  I may misunderstand.

Reply:
Yes, `maxUnavailable` must be zero or else the API rejects the operator's attempt to create the daemonset: `DaemonSet.apps "dns-default" is invalid: spec.updateStrategy.rollingUpdate.maxSurge: Invalid value: intstr.IntOrString{Type:1, IntVal:0, StrVal:"10%"}: may not be set when maxUnavailable is non-zero"`.  That error comes from here: <https://github.com/openshift/kubernetes/blob/54b552033300ab1f94ddf8a9d16032e6619556da/pkg/apis/apps/validation/validation.go#L400-L405>.  The docs are inaccurate.  Someday I might get around to submitting a PR to fix them.

Reply:
For reference, https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#upgrade-and-reconfiguration recommends `maxUnavailable: 10%` for daemonsets for components that have any impact on user workloads and `maxUnavailable: 33%` for daemonsets for components that don't have any such impact.  I am carrying that convention over to `maxSurge` (although maybe the risk of impact from higher percentage values is lower with `maxSurge`—the old pods should only get deleted when the new pods report ready, and if the new pods report ready, then the rollout will proceed anyway and eventually roll out 100% new pods).